### ISSUE TITLE
Type hinting / annotation (PEP 484)?

### ISSUE THREAD

**Issue by f0lie** [nltk/nltk#2020]
Hello folks,

Recently [numpy](https://github.com/numpy/numpy/issues/7370) added some type hinting for their types. Tools like mypy is suppose to make it easier to test large codebases.  There is a recent [pycon](https://www.youtube.com/watch?v=pMgmKJyWKn8) talk about using mypy in the real world. Since nltk is pretty old and large, I figure it might be useful here.

I tried use tools like Monkeytype to generate some stubs for the codebase. But I find it hard to add type hints to a project's code if you aren't familiar with it.

Is there any interest in exploring this route?

Thanks,

**Comment by alvations** [comment]
Yes, type-hints are nice. If you would like to, pick a module or a couple of files that you'll like to add type-hints and create a PR. Someone will review it and take a look to see how it fits in and how it can be extended to more modules/files in `nltk`.

**Comment by f0lie** [comment]
@alvations Can you suggest a few modules?

I am trying to work on [better documentation](https://github.com/python/mypy/issues/5033) for mypy so some concrete projects would be nice.

**Comment by f0lie** [comment]
[](url)

**Comment by f0lie** [comment]
[](url)

**Comment by alvations** [comment]
Maybe you can try `*_score.py` files from https://github.com/nltk/nltk/tree/develop/nltk/translate. They are quite isolated and self-contained. 

Otherwise, a really useful place to have type-hints would be any files/functions/classes in https://github.com/nltk/nltk/tree/develop/nltk/tag

**Comment by f0lie** [comment]
https://github.com/f0lie/nltk/commit/01bb903e00d6aad182fef4c9a96ff8e351c9422a

Here's the first part of type hinting for brill.py. I was surprised how involved the process was to type hint. Fortunately, I am at pycon so I can ask the authors in person for advice. While mypy is awesome, it has a lot of confusing error messages. https://github.com/python/mypy/issues/5055

Note: if you are running mypy right now, you are going to get errors everywhere because most code out there isn't type hinted. Slowly but surely folks.

P.S. this is only supported by Python 3. I could add type comments but that would just mean adding more to the docstring.

**Comment by alvations** [comment]
Thanks @f0lie , the type-hints looks good as a start! Our sincere thanks to the people who's helping at PyCon too. 

I've not tried personally tried mypy, but I guess it does check a single `.py` file too right? 

Unfortunately, support for Python 2 is sort of important still (at least until PyCon 2020), we see people still reporting issues that are Python2 specific. 

We can try adding a branch to pull from yours with the type hints and run test on the CI and try to support the branch with type-hints until we wean off Python 2 but I'm not sure how much support it requires. 

@stevenbird would have a better sense of whether supporting a Python3 only master branch is feasible.

**Comment by iliakur** [comment]
If you plan to support python 2.7 for a long time still, you can also use a cross-compatible syntax for type hints:
https://mypy.readthedocs.io/en/latest/python2.html#type-checking-python-2-code

**Comment by MarioIshac** [comment]
Since Python 2 has reached EOL, would it be worth reexploring this issue?

**Comment by tomaarsen** [comment]
I would love to see Python 3.5+ type hinting now that we no longer support versions that break with this type hinting.

**Comment by eggplants** [comment]
Now, nltk requires is Python 3.7 or later. So we are able to write type annotations mindlessly.

**Comment by Goldziher** [comment]
Please add this

**Comment by KRRT7** [comment]
are type stubs preferred nowadays or directly type hinting the source is ok?
